### PR TITLE
x509_certificate : Add the capability to automatically renew a certificate

### PR DESCRIPTION
**Pull request by julienhuon**
## Description
This change add the capability to automatically renew a certificate in the x509_certificate resource. It adds a new property (renew_before_expiry) wich contains the number of days before expiration. If this limit is reached, the certificate will be renewed.

Merry christmas ! 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
- [X] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/chef/mixin/openssl_helper.rb [hunk 1]
```diff
@@ -401,6 +401,30 @@ def renew_x509_crl(crl, ca_private_key, info)
...
       end
+
+      # Return true if a certificate need to be renewed (or doesn't exist) according to the number
+      # of days before expiration given
+      # @param [string] cert_file path of the cert file or cert content
+      # @param [integer] renew_before_expiry number of days before expiration
+      # @return [true, false]
+      def cert_need_renewall?(cert_file, renew_before_expiry)
+        raise TypeError, "cert_file must be a String object" unless cert_file.is_a?(String)
+        raise TypeError, "renew_before_expiry must be a Integer object" unless renew_before_expiry.is_a?(Integer)
```
Comment (tas50) [inline lib/chef/mixin/openssl_helper.rb]: No need for this check if the property is an Int. Chef will blow up before it gets here.

File: lib/chef/resource/openssl_x509_certificate.rb [hunk 2]
```diff
@@ -109,13 +109,17 @@ class OpensslX509Certificate < Chef::Resource
...
+        default: 5
+
       action :create do
         description "Generate a certificate"
 
-        unless ::File.exist? new_resource.path
+        if cert_need_renewall?(new_resource.path, new_resource.renew_before_expiry)
           converge_by("Create #{@new_resource}") do
             file new_resource.path do
-              action :create_if_missing
```
Comment (tas50) [inline lib/chef/resource/openssl_x509_certificate.rb]: There was a design to not overwrite the cert on disk if present that was part of this design originally. You'll want to change that action based on renew_before_expiry being set.

File: lib/chef/mixin/openssl_helper.rb (hunk 1 above)
Comment (julienhuon) [reply lib/chef/mixin/openssl_helper.rb]: True. I removed it.

File: lib/chef/resource/openssl_x509_certificate.rb (hunk 2 above)
Comment (julienhuon) [reply lib/chef/resource/openssl_x509_certificate.rb]: I assumed that overwrite the cert on disk only if it is expired should be enought. Ok I will overwrite it only if renew_before_expiry is set and the limit reached.

Comment (julienhuon) [reply lib/chef/resource/openssl_x509_certificate.rb]: Corrected. Hope it was what needed.

File: lib/chef/resource/openssl_x509_certificate.rb [hunk 3]
```diff
@@ -109,30 +109,41 @@ class OpensslX509Certificate < Chef::Resource
...
+        file new_resource.path do
+          action :create_if_missing
+          owner new_resource.owner unless new_resource.owner.nil?
+          group new_resource.group unless new_resource.group.nil?
+          mode new_resource.mode unless new_resource.mode.nil?
+          sensitive true
+          content cert.to_pem
+        end
+
+        unless new_resource.renew_before_expiry.nil?
```
Comment (tas50) [inline lib/chef/resource/openssl_x509_certificate.rb]: We should do this all inline so we don't have a resource appear in the log that we skip:

```ruby
if !new_resource.renew_before_expiry.nil? && cert_need_renewall?(new_resource.path, new_resource.renew_before_expiry)
```

File: lib/chef/resource/openssl_x509_certificate.rb [hunk 4]
```diff
@@ -109,30 +109,41 @@ class OpensslX509Certificate < Chef::Resource
       property :ca_key_pass, String,
         description: "The passphrase for CA private key's passphrase."
 
+      property :renew_before_expiry, Integer,
+        description: "The number of days before the expiry. The certificate will be automaticaly renewed when the value is reached."
```
Comment (tas50) [inline lib/chef/resource/openssl_x509_certificate.rb]: we'll want to get an introduced field added here so this shows up in our docs correctly:
```ruby
introduced: 15.7
```

Comment (julienhuon) [reply lib/chef/resource/openssl_x509_certificate.rb]: Done.

File: lib/chef/resource/openssl_x509_certificate.rb (hunk 3 above)
Comment (julienhuon) [reply lib/chef/resource/openssl_x509_certificate.rb]: Done.

**Comment by tas50** [comment]
Looks good. Thanks for the work.

**Pull request merged**